main.js

```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+const core = require('@actions/core');
+const github = require('@actions/github');
+const exec = require('@actions/exec');
+import { chdir } from 'process';
+const glob = require('@actions/glob');
+
+(async () => {
+  try {
+    const credentials = core.getInput('credentials');
+    const charm_path = core.getInput('charm-path');
+    const charmcraft_channel = core.getInput('charmcraft-channel');
+
+    chdir(charm_path);
+
+    await exec.exec('sudo', [
+      'snap',
+      'install',
+      'charmcraft',
+      '--classic',
+      '--channel',
+      charmcraft_channel,
+    ]);
+
+    await exec.exec('charmcraft', ['pack', '--destructive-mode', '-q']);
+
+    const globber = await glob.create('./*.charm');
+    const paths = await globber.glob();
+
+    await Promise.all(
+      paths.map((path) =>
+        exec.exec(
+          'charmcraft',
+          ['upload', '--release=latest/edge', path],
+          { env: { CHARMCRAFT_AUTH: credentials } }
+        )
+      )
+    );
+  } catch (error) {
+    await exec.exec('bash', ['-c', 'cat ~/snap/charmcraft/common/cache/charmcraft/log/*.log']);
+    core.setFailed(error.message);
```

Comment:
Not sure if this should catch the error or not, but I ran this with an invalid `path` and the action [finished successfully in github](https://github.com/canonical/kfp-operators/runs/4373543814?check_suite_focus=true) (:heavy_check_mark:) but actually failed with the shown message.  It would be good if that situation:
* failed loudly rather than silently
* had an error message that stated what happened (it was easy for me to see my typo but in general this might be hard to debug)

Reply:
I think the silent error is due to this bug: https://github.com/canonical/charmcraft/issues/478

Not much we can do there that isn't a fragile hack (we could e.g. parse the output and try to guess what constitutes an error, but that's very fragile)

Reply:
I updated the error handling to always upload all charmcraft logs as an artifact and skip the progress logging, so hopefully it's pretty easy to see where something has failed now.